#188 <Coupon> 쿠폰 발행한 유저의 데이터베이스에 요청한 쿠폰 데이터 저장

seungg8361 opened the pull request:
- Redis에 먼저 데이터를 저장한 뒤 decrement로 원자적 감소
- 만약 수량이 없으면 FAILED 로 반환
- SUCCESS로 전달된 요청의 데이터만 p_user_coupon 데이터베이스에 저장

## 개요
#188 Coupon 쿠폰 발행한 유저의 데이터베이스에 요청한 쿠폰 데이터 저장

## 작업 내용
### 변경 사항
- Redis에 먼저 데이터를 저장한 뒤 decrement로 원자적 감소
- 만약 수량이 없으면 FAILED 로 반환
- SUCCESS로 전달된 요청의 데이터만 p_user_coupon 데이터베이스에 저장
### 변경 이유

### 추가 설명

## 리뷰 요구사항

#### 연관된 이슈
closed #<이슈 번호>

## PR 유형

이 PR은 어떤 변경 사항을 포함하고 있나요?

- [ ] 새로운 기능 추가
- [ ] 버그 수정
- [ ] UI 디자인 변경 (CSS 등)
- [ ] 코드에 영향을 주지 않는 변경사항 (오타 수정, 탭 사이즈 변경, 변수명 변경 등)
- [ ] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 또는 폴더명 수정
- [ ] 파일 또는 폴더 삭제
- [ ] 배포 준비, PR 관련 사항

## Check List

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다. (버그 수정/기능에 대한 테스트)
- [ ] CI/CD 파이프라인을 통과했습니다.

seungg8361 committed:
feat : #188 Coupon 쿠폰 발행한 유저의 데이터베이스에 요청한 쿠폰 데이터 저장

- Redis에 먼저 데이터를 저장한 뒤 decrement로 원자적 감소
- 만약 수량이 없으면 FAILED 로 반환
- SUCCESS로 전달된 요청의 데이터만 p_user_coupon 데이터베이스에 저장

seungg8361 committed:
refactor : #188 Coupon 쿠폰 발행한 유저의 데이터베이스에 요청한 쿠폰 데이터 저장

- Quantity 타입을 Long으로 변경
- redis로 quantity 수량 조절이 끝나면 기존 데이터베이스의 quantity 업데이트
- 이후 TTL을 적용하지 않고, redis key를 삭제해서 메모리, 데이터 유실을 최소화

jjsh0208 reviewed (approved):
고생하셨습니다! 메시지 안에 enum 참고할게용

seungg8361 committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into feature/#188

seungg8361 committed:
fix : #188 Coupon 쿠폰 발행한 유저 기능 수정

- 쿠폰 선착순 기능 전면 수정사항과 쿠폰을 포함한 데이터 조회 시, redis에 실시간으로 정보가 업데이트 되지 않아서 최신 사항을 조회하지 못하는 오류를 수정했습니다.
- 마일리지 및 쿠폰에 새로운 데이터가 추가될 때마다 기존 redis의 조회 캐싱을 삭제합니다.

seungg8361 committed:
refactor : #188 Coupon 쿠폰 발행한 유저 기능

- expired_at을 현재 시간보다 10분 늦게 설정
- Coupon 도메인에 discount, expired_at 추가
- 발행에 설공 or 샐패의 유무를 알 수 있게 아주 간단한 API 추가
